apstra/api_custom_collector.go

```diff
@@ -0,0 +1,257 @@
+		}
+		o.OsFamily = append(o.OsFamily, variant)
+	}
+
+	return nil
+}
+
+func (o *CollectorPlatform) MarshalJSON() ([]byte, error) {
+	var raw struct {
+		OsType    string `json:"os_type"`
+		OsVersion string `json:"os_version"`
+		OsFamily  string `json:"family"`
+		Model     string `json:"model"`
+	}
+	raw.OsType = o.OsType.String()
+	raw.OsVersion = o.OsVersion.String()
+	raw.Model = o.Model
+	raw.OsFamily = o.OsFamily[0].String()
+	for _, v := range o.OsFamily[1:] {
+		raw.OsFamily = raw.OsFamily + "," + v.String()
+	}
+	return json.Marshal(raw)
+}
+
+type Query struct {
+	Accessors map[string]string `json:"accessors"`
+	Keys      map[string]string `json:"keys"`
+	Value     string            `json:"value"`
+	Filter    string            `json:"filter"`
+}
+type Collector struct {
+	ServiceName             string
+	Platform                CollectorPlatform `json:"platform"`
+	SourceType              string            `json:"source_type"`
+	Cli                     string            `json:"cli"`
+	Query                   Query             `json:"query"`
+	RelaxedSchemaValidation bool              `json:"relaxed_schema_validation"`
+}
+
+// GetAllCollectors gets all the Collectors for all services
+func (o *Client) GetAllCollectors(ctx context.Context) ([]Collector, error) {
+	var response struct {
+		Items map[string]interface{} `json:"items"`
+	}
+	var collectors []Collector
+	// First Reach out to /collectors , we are interested really only in the keys, so that we can pull the collectors
+	err := o.talkToApstra(ctx, &talkToApstraIn{
+		method:      http.MethodGet,
+		urlStr:      fmt.Sprintf(apiUrlCollectors),
+		apiResponse: &response,
+	})
+	if err != nil {
+		return nil, convertTtaeToAceWherePossible(err)
+	}
+
+	for k := range response.Items {
+		cs, err := o.GetCollectorsByServiceName(ctx, k)
+		if err != nil {
+			return nil, err
+		}
+		for _, v := range cs {
+			v.ServiceName = k
+			collectors = append(collectors, v)
+		}
+	}
+	return collectors, nil
+}
+
+// GetCollectorsByServiceName gets all the Collectors that correspond to a particular service
+func (o *Client) GetCollectorsByServiceName(ctx context.Context, name string) ([]Collector, error) {
+	var ace ClientErr
+	var Response struct {
+		Items []Collector `json:"items"`
+	}
+	err := o.talkToApstra(ctx, &talkToApstraIn{
+		method:      http.MethodGet,
+		urlStr:      fmt.Sprintf(apiUrlCollectorsByServiceName, name),
+		apiResponse: &Response,
+	})
+
+	if err != nil {
+		err = convertTtaeToAceWherePossible(err)
+		if errors.As(err, &ace) && ace.Type() == ErrNotfound {
+			return Response.Items, nil
+		}
+		return nil, err
+	}
+
+	for i := range Response.Items {
+		Response.Items[i].ServiceName = name
+	}
+	return Response.Items, nil
+}
+
+// CreateCollector creates a collector
+func (o *Client) CreateCollector(ctx context.Context, in *Collector) error {
+	// Check if this is the first collector for that service name
+	cs, err := o.GetCollectorsByServiceName(ctx, in.ServiceName)
```

Comment:
Is this the thing we talked about trying to avoid the race condition here by starting with a POST? If that fails because of existing service name, absorb the error and try PATCH.

Did that work out?

I guess it still invites a race condition (existing object deleted between POST and PATCH), but this option feels slightly safer.

Reply:
yes, this pattern works, but as you have noted further down, I should probably do some locking

Reply:
I think I wasn't clear. My feeling is that starting with `POST` (and handling the error) feels better than starting with `GET` and using that to determine whether you should `POST` or `PATCH`.

My reasoning:
- `GET` first opens the door to race between two invocations of `Create()` ... this feels like a more common issue.
- `POST` first opens the door to race between `Create()` and the last `Delete()` ... this feels like a less common issue.
- `GET` first always requires two transactions, where `POST` first *might* get the job done in one.

This is all assuming `POST` first is practicable. We talked about it, but I don't remember whether it worked out.

Reply:
Fair point, removed the get, did post->patch

Reply:
I think this function still needs some attention.

There's some dead code, the comments seems to make false assertions, and I think we will wind up doing both POST and PATCH right now.